tools/uebuild/uebuild.nim

```diff
@@ -363,19 +361,28 @@ proc buildNim(projectDir, projectName, os, cpu, uePlatform: string, isEditorBuil
         withDir nimOutDir:
           exec "nimble -y cpp \"" & rootFile & '"'
 
-      for file in walkDirRec(nimcacheDir, {pcFile}):
-        if file.endsWith(".h") and not expectedFilenames.contains(extractFilename(file)):
-          let cppFile = file.changeFileExt("cpp")
-          let cppFilename = cppFile.extractFilename()
-          removeFile file
-          removeFile cppFile
-          removeFile targetDir / file.extractFilename()
-          removeFile targetDir / "Public" / file.extractFilename()
-          removeFile targetDir / "Private" / file.extractFilename()
-          removeFile targetDir / cppFilename
-          removeFile targetDir / "Private" / cppFilename
-        elif file.endsWith(".cpp"):
+      let jsonfile = nimcacheDir / projectName & "Root.json"
+      if fileExists(jsonfile):
+        let data = json.parseFile(jsonfile)
+        for cfileEntry in data["compile"]:
+          let file = cfileEntry[0].str
           processFile(file, moduleName, targetDir, nimblePackageName)
```

Comment:
There should be code that removes redundant files (files not present in "compile" array) from the targetDir. Otherwise, we still have to run `clean` to remove them. So we should walk through files in `targetDir/"Private"`, for each file that ends with `.cpp` check if it's in "compile" array, and if it's not, delete the file and its .h counterpart (that may or may not exist).